marlowe-dashboard-client/src/MainFrame/State.purs

```diff
@@ -213,123 +246,158 @@ handleAction AddNewWallet = do
+      contractInstanceId = view _contractInstanceId walletDetails
+    rdCompanionContractState <- marloweGetWalletCompanionContractObservableState contractInstanceId
+    rdContractInstanceIds <- getWalletContractInstances wallet
+    case rdAssets, rdCompanionContractState, rdContractInstanceIds of
+      Success assets, Success companionContractState, Success contractInstanceIds -> do
+        let
+          updatedWalletDetails = set _assets assets walletDetails
+        -- subscribe to the wallet to get updates about balance changes
+        subscribeToWallet wallet
+        -- FIXME: make sure all contracts are started and get latest contract state
+        -- FIXME: subscribe to each contract to get updates about their state
+        slot <- liftEffect currentSlot
+        let
+          contracts = dummyContracts slot
+        modify_
+          $ set _subState (Right $ Play.mkInitialState updatedWalletDetails contracts timezoneOffset)
+          <<< set (_pickupState <<< _card) Nothing
+        -- save the wallet details to local storage, so the user will automatically pick it up when they return
+        liftEffect $ setItem walletDetailsLocalStorageKey $ encodeJSON updatedWalletDetails
+      -- TODO: show more specific errors for the different cases
+      Failure err, _, _ -> networkErrorToast err
+      _, _, _ -> pure unit
+  Pickup.GenerateNewWallet -> do
+    remoteDataWalletInfo <- createWallet
+    assign (_newWalletDetails <<< _remoteDataWalletInfo) remoteDataWalletInfo
+    let
+      networkErrorToast = \_ -> addToast $ connectivityErrorToast "Couldn't generate wallet"
+    case remoteDataWalletInfo of
+      Success (WalletInfo { wallet }) -> do
+        remoteDataAssets <- getWalletTotalFunds wallet
+        assign (_newWalletDetails <<< _remoteDataAssets) remoteDataAssets
+        remoteDataContractInstanceId <- marloweCreateWalletCompanionContract wallet
+        case remoteDataContractInstanceId of
+          Success contractInstanceId -> do
+            let
+              contractInstanceIdString = UUID.toString $ unwrap contractInstanceId
+            modify_
+              $ set (_newWalletDetails <<< _contractInstanceIdString) contractInstanceIdString
+              <<< set (_pickupState <<< _card) (Just (PickupNewWalletCard WalletGenerated))
+          Failure err -> networkErrorToast err
+          _ -> pure unit
+      Failure err -> networkErrorToast err
+      _ -> pure unit
+  Pickup.SetPickupWalletString string -> do
+    assign (_pickupState <<< _pickupWalletString) string
+    wallets <- use _wallets
+    -- first check for a matching nickname in the wallet library
+    case lookup string wallets of
+      Just walletDetails -> assign (_pickupState <<< _card) $ Just $ PickupWalletCard walletDetails
+      -- then check for a matching ID in the wallet library
+      Nothing -> case findMin $ filter (\walletDetails -> UUID.toString (unwrap (view _contractInstanceId walletDetails)) == string) wallets of
```

Comment:
if we are using lenses, maybe this could work `view (_Newtype <<< _contractInstanceId)` so we don't have to unwrap later on.

Reply:
good point 👍